fedimint-atomic-broadcast/src/lib.rs

```diff
@@ -0,0 +1,246 @@
+//! all correct nodes if a correct peer accepts it. Those two guarantees are
+//! sufficient to build consistent replicated state machines like Fedimint
+//! Consensus on top of the broadcast. Such a state machine has to accept an
+//! item if it changes the machines state and should discard it otherwise. Let
+//! us consider the case of an ecash note being double spend by the items
+//! A and B while one peer is offline. First, item A is ordered and all correct
+//! peers include the note as spent in their state. Therfore they also accept
+//! the item A. Then, item B is ordered and all correct nodes notice the double
+//! spend and make no changes to their state. Now they can safely discard the
+//! item B as it did not cause a state transition. When the session completes
+//! only item A is part of the corresponding block. When the offline peer comes
+//! back online it downloads the block. Therefore the recovering peer will only
+//! see Item A but arrives at the same state as its peers at the end of the
+//! session regardless. However, it did so by processing one less ordered item
+//! and without realizing that a double spend had occured.
+//!
+//! Since a state machine may process only a subsequence of a sessions items we
+//! can not simply rely on the last processed items index when recovering from a
+//! crash mid session in order to know how many items from the session have
+//! already been applied to our state. Furthermore, the broadcast instance does
+//! not save the decisions for items processed before the crash to disk so
+//! Fedimint Consensus either has to keep track of the decisions on disk or
+//! recompute them. I think the simplest solution to this is to use a single
+//! database transaction for the entire session for all consensus critical state
+//! and maintain the index of the current session as part of the state machine.
+//! If we crash mid session all intermediate state changes are lost and we
+//! simply process the returned ordered items again without any recovery
+//! specific logic. At the end of a session all state machines will arrive at
+//! the same state regardless of the exact number of ordered items they
+//! processed to arrive there. Hence, when we receive the first ordered item of
+//! the next session which indicates that the current session is complete we can
+//! bump the index of current session and finally commit the transaction.
+
+mod broadcast;
+mod conversion;
+mod data_provider;
+mod db;
+mod finalization_handler;
+mod keychain;
+mod network;
+mod session;
+mod spawner;
+
+use bitcoin::merkle_tree;
+use bitcoin_hashes::{sha256, Hash};
+/// This function runs the broadcast until a shutdown is intiated either via the
+/// shutdown sender or the [OrderedItem] receiver is dropped.
+pub use broadcast::run;
+use fedimint_core::encoding::{Decodable, Encodable};
+use fedimint_core::PeerId;
+/// This keychain implements naive threshold schnorr signatures over secp256k1.
```

Comment:
Probably better to move these comments to the item being imported themselves.

Reply:
I put it there such that all public documentation is in one file